src/xenia/kernel/xam/xam_net.cc

```diff
@@ -1031,6 +1008,28 @@ void NetDll_WSASetLastError_entry(dword_t error_code) {
 }
 DECLARE_XAM_EXPORT1(NetDll_WSASetLastError, kNetworking, kImplemented);
 
+dword_result_t NetDll_getsockname_entry(dword_t caller, dword_t socket_handle,
+                                        lpvoid_t buf_ptr, lpdword_t len_ptr) {
+  auto socket =
+      kernel_state()->object_table()->LookupObject<XSocket>(socket_handle);
+  if (!socket) {
+    XThread::SetLastError(uint32_t(X_WSAError::X_WSAENOTSOCK));
+    return -1;
+  }
+
+  int buffer_len = *len_ptr;
```

Comment:
According to the [MSDN documentation for `getsockname`](https://learn.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-getsockname):

> WSAEFAULT
> The name or the namelen parameter is not a valid part of the user address space, or the namelen parameter is too small.

I'm not sure if that's the case on the Xbox 360 exactly, but I think `WSAEFAULT` should be returned at least for nullptr `buf_ptr` or `len_ptr`, though generally since it's a kernel function it can have more thorough validation of pointers against the memory page tables.

However, a crash may be fine too since games aren't supposed to be doing that anyway. I was just looking if nullptr inputs needed to be handled specially, but looks like they don't need to.

Reply:
Some validation, however, may be useful for consistency of error codes. On Windows, errors related to `namelen` are reported as `WSAEFAULT`, but on POSIX, they produce the `EINVAL` errno.

Also, there may be differences between the behavior in case the length is too small, but those need to be verified. On POSIX, the result is implicitly truncated if the length is too small, and the function probably succeeds (it only returns `EINVAL` if the length is invalid, like negative). On Windows, I'm not sure, but I can't find any info on MSDN about truncation, and the conditions of `WSAEFAULT` include "the `namelen` parameter is too small". However, I think if the function fails, it shouldn't write anything to any outputs (and `WSAEFAULT` doesn't distinguish between too small `namelen` and invalid `namelen`, so there's no fully reliable way to check if the function has actually written the full `namelen` needed — if the buffer is just too small, the function should probably succeed anyway) — but need to verify.

Though all of this probably should be handled on the `XSocket` side so it provides a more or less unified API between platforms, not in the kernel guest entry points.